diff -u -r --ignore-space-change -N ldmud-3.2.6/src/Makefile.in ldmud-3.2.6-ip6/src/Makefile.in
--- ldmud-3.2.6/src/Makefile.in	Tue Mar 30 18:28:11 1999
+++ ldmud-3.2.6-ip6/src/Makefile.in	Thu Jul 22 06:29:15 1999
@@ -58,9 +58,9 @@
 #
 MPATH=-DMUD_LIB='"$(MUD_LIB)"' -DBINDIR='"$(BINDIR)"'
 #
-CFLAGS= @EXTRA_CFLAGS@ $(OPTIMIZE) $(DEBUG) -DMALLOC_$(MALLOC) $(WARN) $(MPATH) $(PROFIL)
+CFLAGS= -I/usr/inet6/include @EXTRA_CFLAGS@ $(OPTIMIZE) $(DEBUG) -DMALLOC_$(MALLOC) $(WARN) $(MPATH) $(PROFIL)
 #
-LIBS=@LIBS@
+LIBS=@LIBS@ -linet6
 #
 LDFLAGS=@LDFLAGS@
 #
@@ -75,7 +75,7 @@
       port.c ptrtable.c \
       random.c regexp.c rxcache.c simulate.c simul_efun.c stralloc.c \
       sprintf.c swap.c wiz_list.c $(MALLOC).c
-OBJ=lang.o lex.o main.o interpret.o simulate.o object.o backend.o array.o\
+OBJ=mudipv6.o lang.o lex.o main.o interpret.o simulate.o object.o backend.o array.o\
     comm.o ed.o regexp.o mapping.o wiz_list.o swap.o $(MALLOC).o\
     call_out.o otable.o dumpstat.o stralloc.o hash.o port.o\
     access_check.o parse_old.o parse.o simul_efun.o sprintf.o\
diff -u -r --ignore-space-change -N ldmud-3.2.6/src/comm.c ldmud-3.2.6-ip6/src/comm.c
--- ldmud-3.2.6/src/comm.c	Tue Mar 30 18:28:11 1999
+++ ldmud-3.2.6-ip6/src/comm.c	Tue Jul 20 07:26:44 1999
@@ -526,10 +526,10 @@
      */
     if (udp_port != -1)
     {
-        host_ip_addr.sin_addr.s_addr = INADDR_ANY;
+        host_ip_addr.sin_addr = in6addr_any;
         host_ip_addr.sin_port = htons((u_short)udp_port);
         debug_message("UDP recv-socket requested for port: %d\n", udp_port);
-        udp_s = socket(host_ip_addr.sin_family, SOCK_DGRAM, 0);
+        udp_s = socket(AF_INET6, SOCK_DGRAM, 0);
         if (udp_s == -1)
         {
             perror("socket(udp_socket)");
@@ -667,8 +667,8 @@
             /* Real port number */

             host_ip_addr.sin_port = htons((u_short)PNUMBER);
-            host_ip_addr.sin_addr.s_addr = INADDR_ANY;
-            MUDSOCKET = socket(host_ip_addr.sin_family, SOCK_STREAM, 0);
+            host_ip_addr.sin_addr = in6addr_any;
+            MUDSOCKET = socket(AF_INET6, SOCK_STREAM, 0);
             if ((int)MUDSOCKET == -1) {
                 perror("socket");
                 exit(1);
@@ -1659,7 +1659,7 @@
                 trace_level = 0;
                 udp_buf[sizeof(udp_buf) - 1] = '\0';
                 udp_buf[cnt] = '\0';
-                st = inet_ntoa(addr.sin_addr);
+                st = inet6_ntoa(addr.sin_addr);
                 push_string_malloced(st);
                 push_string_malloced(udp_buf);
                 push_number(ntohs(addr.sin_port));
@@ -4156,7 +4156,7 @@
     /* Return the IP address as string.
      */

-    str = string_copy(inet_ntoa(ip->addr.sin_addr));
+    str = string_copy(inet6_ntoa(ip->addr.sin_addr));
     if (!str)
     {
         inter_sp = sp - 1;
@@ -4200,7 +4200,7 @@
 {
     char buf[20];

-    sprintf(buf, "\"%s\"", inet_ntoa(host_ip_number));
+    sprintf(buf, "\"%s\"", inet6_ntoa(host_ip_number));
     return string_copy(buf);
 }

@@ -4746,8 +4746,8 @@

         if (sscanf(to_host, "%d.%d.%d.%d", &ip1, &ip2, &ip3, &ip4) == 4)
         {
-            name.sin_addr.s_addr = inet_addr(to_host);
-            name.sin_family = AF_INET;
+            name.sin_addr = inet6_addr(to_host);
+            name.sin_family = AF_INET6;
         }
         else
         {
@@ -4755,7 +4755,7 @@
             if (hp == 0)
                 break;
             memcpy(&name.sin_addr, hp->h_addr, hp->h_length);
-            name.sin_family = AF_INET;
+            name.sin_family = AF_INET6;
         }
         name.sin_port = htons(to_port);

diff -u -r --ignore-space-change -N ldmud-3.2.6/src/comm.h ldmud-3.2.6-ip6/src/comm.h
--- ldmud-3.2.6/src/comm.h	Wed Mar 10 16:57:45 1999
+++ ldmud-3.2.6-ip6/src/comm.h	Tue Jul 20 06:14:57 1999
@@ -43,6 +42,7 @@

 #define MAX_TEXT  2048

+#include "mudipv6.h"

 /* --- Types --- */

diff -u -r --ignore-space-change -N ldmud-3.2.6/src/mudipv6.c ldmud-3.2.6-ip6/src/mudipv6.c
--- ldmud-3.2.6/src/mudipv6.c	Wed Dec 31 16:00:00 1969
+++ ldmud-3.2.6-ip6/src/mudipv6.c	Tue Jul 20 06:09:22 1999
@@ -0,0 +1,21 @@
+#include <netinet/in.h>
+
+#include "mudipv6.h"
+
+char *inet6_ntoa(struct in6_addr in)
+{
+	static char str[4097];
+
+	inet_ntop(AF_INET6, &in, str, 4096);
+	return str;
+}
+
+
+struct in6_addr inet6_addr(char *to_host)
+{
+	struct in6_addr addr;
+
+	inet_pton(AF_INET6, to_host, &addr);
+	return addr;
+}
+
diff -u -r --ignore-space-change -N ldmud-3.2.6/src/mudipv6.h ldmud-3.2.6-ip6/src/mudipv6.h
--- ldmud-3.2.6/src/mudipv6.h	Wed Dec 31 16:00:00 1969
+++ ldmud-3.2.6-ip6/src/mudipv6.h	Tue Jul 20 06:09:22 1999
@@ -0,0 +1,17 @@
+#ifndef _driverv6_
+
+	#define _driverv6_
+	#define sockaddr_in sockaddr_in6
+
+	#define sin_port sin6_port
+	#define sin_addr sin6_addr
+	#define sin_family sin6_family
+	#define s_addr s6_addr
+	#define in_addr in6_addr
+
+char *inet6_ntoa(struct in6_addr in);
+
+struct in6_addr inet6_addr(char *to_host);
+
+#endif
+
