diff -u -r MG-326-orig/secure/inetd.c MG-326/secure/inetd.c
--- MG-326-orig/secure/inetd.c	Sat Jun 12 14:40:38 1999
+++ MG-326/secure/inetd.c	Wed May  9 16:28:52 2001
@@ -105,7 +105,7 @@
 private int packet_id;

 void set_host_list();
-varargs int send_udp(string mudname, mapping data, int expect_reply);
+varargs string send_udp(string mudname, mapping data, int expect_reply);

 #define DEBUG(msg)	if (find_player("jof"))\
 			tell_object(find_player("jof"), msg);
diff -u -r MG-326-orig/secure/login.c MG-326/secure/login.c
--- MG-326-orig/secure/login.c	Tue Jan 11 15:03:32 2000
+++ MG-326/secure/login.c	Wed May  9 16:26:42 2001
@@ -622,7 +622,7 @@
 void _telnet_neg(mixed cmd,mixed opt,mixed args)
 {
   if (opt==34 && cmd==251)
-    binary_message(({255,250,34,1,1,255,240}));
+    binary_message(({255,250,34,1,1,255,240}), 0);
 }

 int BLA(string s)
diff -u -r MG-326-orig/secure/simul_efun.c MG-326/secure/simul_efun.c
--- MG-326-orig/secure/simul_efun.c	Sun Feb 27 02:47:19 2000
+++ MG-326/secure/simul_efun.c	Wed May  9 16:37:00 2001
@@ -48,7 +48,7 @@
 varargs int log_file(string file, string txt, int size_to_break);
 int query_wiz_level(mixed player);
 mixed *remove_alist(mixed key, mixed *alist);
-nomask int snoop(object me, object you);
+varargs nomask int snoop(object me, object you);
 nomask object query_snoop(object who);
 string *get_dir_ext(string path);
 string *new_explode(string str, string del);
@@ -126,11 +126,12 @@
   return ob->_query_my_ip()||efun::query_ip_number(ob);
 }

-string query_ip_name(object ob)
+varargs string query_ip_name(object ob)
 {
   string ipnum;
   mixed ipname;

+  if (!ob) ob = this_interactive();
   ipnum=query_ip_number(ob);
   ipname=ip_map[ipnum,0];
   if (stringp(ipname))
@@ -293,7 +294,7 @@
   return str[5..11]+str[18..];
 }

-nomask int snoop( object me, object you )
+varargs nomask int snoop( object me, object you )
 {
     int ret;
     object snooper0, snooper, snooper2, snooper3;
@@ -1097,7 +1098,7 @@

 // * wizlist ausgeben

-void wizlist(string name) {
+varargs void wizlist(string name) {
   int i, pos, total_cmd;
   int *cmds;
   mixed *a;
@@ -1798,20 +1799,20 @@
   return;
 }

-void notify_fail(string str)
+int notify_fail(string str)
 {
-  if (!TP) return;
+  if (!TP) return 0;
   TP->SetProp(P_ACTUAL_NOTIFY_FAIL,previous_object());
   return efun::notify_fail(str);
 }

-void _notify_fail(string str)
+int _notify_fail(string str)
 {
   object o;

-  if (!TP) return;
+  if (!TP) return 0;
   if ((o=TP->QueryProp(P_ACTUAL_NOTIFY_FAIL)) && o!=previous_object())
-    return;
+    return 0;
   TP->SetProp(P_ACTUAL_NOTIFY_FAIL,previous_object());
   return efun::notify_fail(str);
 }
@@ -2041,4 +2042,4 @@
    // Lichtberechnung findet eigentlich in der Mudlib statt.
    return ob->QueryProp(P_INT_LIGHT);
 }
-
\ No newline at end of file
+
diff -u -r MG-326-orig/secure/spare_simul_efun.c MG-326/secure/spare_simul_efun.c
--- MG-326-orig/secure/spare_simul_efun.c	Sun Jan  9 15:20:12 2000
+++ MG-326/secure/spare_simul_efun.c	Wed May  9 16:36:52 2001
@@ -48,7 +48,7 @@
 varargs int log_file(string file, string txt, int size_to_break);
 int query_wiz_level(mixed player);
 mixed *remove_alist(mixed key, mixed *alist);
-nomask int snoop(object me, object you);
+varargs nomask int snoop(object me, object you);
 nomask object query_snoop(object who);
 string *get_dir_ext(string path);
 string *new_explode(string str, string del);
@@ -126,11 +126,12 @@
   return ob->_query_my_ip()||efun::query_ip_number(ob);
 }

-string query_ip_name(object ob)
+varargs string query_ip_name(object ob)
 {
   string ipnum;
   mixed ipname;

+  if (!ob) ob = this_interactive();
   ipnum=query_ip_number(ob);
   ipname=ip_map[ipnum,0];
   if (stringp(ipname))
@@ -293,7 +294,7 @@
   return str[5..11]+str[18..];
 }

-nomask int snoop( object me, object you )
+varargs nomask int snoop( object me, object you )
 {
     int ret;
     object snooper0, snooper, snooper2, snooper3;
@@ -1097,7 +1098,7 @@

 // * wizlist ausgeben

-void wizlist(string name) {
+varargs void wizlist(string name) {
   int i, pos, total_cmd;
   int *cmds;
   mixed *a;
@@ -1798,20 +1799,20 @@
   return;
 }

-void notify_fail(string str)
+int notify_fail(string str)
 {
-  if (!TP) return;
+  if (!TP) return 0;
   TP->SetProp(P_ACTUAL_NOTIFY_FAIL,previous_object());
   return efun::notify_fail(str);
 }

-void _notify_fail(string str)
+int _notify_fail(string str)
 {
   object o;

-  if (!TP) return;
+  if (!TP) return 0;
   if ((o=TP->QueryProp(P_ACTUAL_NOTIFY_FAIL)) && o!=previous_object())
-    return;
+    return 0;
   TP->SetProp(P_ACTUAL_NOTIFY_FAIL,previous_object());
   return efun::notify_fail(str);
 }
@@ -2041,4 +2042,4 @@
    // Lichtberechnung findet eigentlich in der Mudlib statt.
    return ob->QueryProp(P_INT_LIGHT);
 }
-
\ No newline at end of file
+
diff -u -r MG-326-orig/std/living/skills.c MG-326/std/living/skills.c
--- MG-326-orig/std/living/skills.c	Thu Jan 13 09:17:58 2000
+++ MG-326/std/living/skills.c	Wed May  9 16:40:56 2001
@@ -465,7 +465,7 @@
   // direkt aufgerufen werden... Hier hat es einiges vereinfacht...
 }

-varargs void LearnSkill(string sname, int add, int diff) {
+varargs int LearnSkill(string sname, int add, int diff) {
   mapping res;
   string skill2,gilde;
   int val;
@@ -486,6 +486,7 @@
   ModifySkill(sname,val,diff,gilde);
   if (mappingp(res) && skill2=res[SI_INHERIT])
     LearnSkill(skill2,add/3,diff);
+  return 0;
 }

 varargs int return0() {return 0;}
diff -u -r MG-326-orig/std/player/channel.c MG-326/std/player/channel.c
--- MG-326-orig/std/player/channel.c	Tue Jan 25 08:11:10 2000
+++ MG-326/std/player/channel.c	Wed May  9 16:25:55 2001
@@ -285,7 +285,7 @@
     }
     case '*':
     {
-      mixed hist; int amount;
+      mixed hist; int amount; string txt;
       if(!pointerp(hist = CHMASTER->history(ch, this_object()))
 		      || !sizeof(hist))
       {
@@ -390,7 +390,7 @@
   }
   if(sscanf(args, "beschreibung %s %s", n, descr) == 2)
   {
-    mixed ch;
+    mixed ch, cn;
     cn = CHMASTER->find(n, this_object());
     if(!cn || pointerp(cn))
       return (notify_fail("Die Ebene '"+n+"' existiert nicht oder die Angabe "
@@ -475,6 +475,8 @@
                             +"' eingestellt.\n"
                             : "Es ist keine Standardebene eingestellt.\n")),0);
     else
+    {
+      mixed tmp;
       if(pointerp(tmp = CHMASTER->find(pa[1], this_object())))
         return (notify_fail("Das war keine eindeutige Angabe! "
                             "Folgende Ebenen passen:\n"
@@ -487,4 +489,5 @@
           SetProp(P_STD_CHANNEL, tmp);
           return 1;
         }
+    }
 }
diff -u -r MG-326-orig/std/shells/filesys/make.c MG-326/std/shells/filesys/make.c
--- MG-326-orig/std/shells/filesys/make.c	Tue Jan 25 09:34:53 2000
+++ MG-326/std/shells/filesys/make.c	Wed May  9 16:38:53 2001
@@ -79,7 +79,7 @@
   return 1;
 }

-varargs private void _make(string file, int flags);
+varargs private int _make(string file, int flags);

 // _instance_upd() -- update all instances found for a class
 private void _instance_upd(string file, int flags, mixed o, int instances)
diff -u -r MG-326-orig/std/shells/filesys/telnetneg.c MG-326/std/shells/filesys/telnetneg.c
--- MG-326-orig/std/shells/filesys/telnetneg.c	Wed Mar 24 11:19:11 1999
+++ MG-326/std/shells/filesys/telnetneg.c	Wed May  9 16:30:26 2001
@@ -11,10 +11,11 @@

 #define NEED_PROTOTYPES

+inherit "/std/thing/properties";
+
 #include <properties.h>
 #include <telnet.h>

-inherit "/std/thing/properties";

 static mapping TN;
 static string *Terminals;
diff -u -r MG-326-orig/std/shells/magier.c MG-326/std/shells/magier.c
--- MG-326-orig/std/shells/magier.c	Fri Feb 11 07:56:28 2000
+++ MG-326/std/shells/magier.c	Wed May  9 16:34:30 2001
@@ -995,7 +995,7 @@
   return 1;
 }

-void quit();
+int quit();

 void reset()
 {
diff -u -r MG-326-orig/std/thing/description.c MG-326/std/thing/description.c
--- MG-326-orig/std/thing/description.c	Fri Feb 25 08:10:10 2000
+++ MG-326/std/thing/description.c	Wed May  9 16:23:25 2001
@@ -239,7 +239,7 @@
 varargs string name( int casus, int demon )
 {
     string adj;
-    mixed *sh, last;
+    mixed sh, last;
     int noart, plural;

     noart = !(ME->QueryProp(P_ARTICLE));
diff -u -r MG-326-orig/sys/daemon/channel.h MG-326/sys/daemon/channel.h
--- MG-326-orig/sys/daemon/channel.h	Wed Feb  9 08:49:58 2000
+++ MG-326/sys/daemon/channel.h	Wed May  9 16:24:03 2001
@@ -40,7 +40,7 @@
 #define I_NAME                  4
 #define I_TIMEOUT               5

-#endif __DAEMON_CHANNEL_H__
+#endif /* __DAEMON_CHANNEL_H__ */

 // prototypes
 #ifdef NEED_PROTOTYPES
@@ -54,6 +54,6 @@
 varargs int send(string ch, mixed pl, string msg, int type);
 mixed list(mixed pl);

-#endif __CHANNEL_H_PROTO__
+#endif /* __CHANNEL_H_PROTO__ */

-#endif NEED_PROTOTYPES
+#endif /* NEED_PROTOTYPES */
diff -u -r MG-326-orig/sys/daemon/html.h MG-326/sys/daemon/html.h
--- MG-326-orig/sys/daemon/html.h	Thu May 27 04:22:46 1999
+++ MG-326/sys/daemon/html.h	Wed May  9 16:16:25 2001
@@ -40,7 +40,7 @@
 #define H_PRE		  "pre"		// preformatted text
 #define H_BREAK		"br"		// force line break;

-#endif __HTML_H__
+#endif /* __HTML_H__ */



diff -u -r MG-326-orig/sys/new_skills.h MG-326/sys/new_skills.h
--- MG-326-orig/sys/new_skills.h	Wed Feb  9 08:54:10 2000
+++ MG-326/sys/new_skills.h	Wed May  9 16:38:05 2001
@@ -262,7 +262,7 @@
                                   int duration, mixed fun);
 mapping QuerySkill(string sname);
 varargs void ModifySkill(string sname, mixed val, int diff, string gilde);
-varargs void LearnSkill(string sname, int add, int diff);
+varargs int LearnSkill(string sname, int add, int diff);
 varargs int UseSpell(string str, string spell);
 varargs mixed UseSkill(string skill, mapping args);
 varargs int QuerySkillAbility(mixed sname, mixed argl);
